### ISSUE TITLE
Review package examples from manual

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#253291]
## Issue description

We teach packaging using examples, but i think these examples don't follow latest best practices.

https://nixos.org/manual/nixpkgs/stable/#chap-quick-start

One thing is that in `src` it often uses `sha256 =` instead of `hash =` which is used in the Fetching Sources section. The manual should provide consistent information.

- [ ] https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/misc/hello/default.nix
- [ ] https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/archivers/cpio/default.nix
- [ ] https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/gmp/5.1.x.nix
- [ ] https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/newsreaders/pan/default.nix
- [ ] https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/http/apache-httpd/2.4.nix
- [ ] https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/browsers/firefox/common.nix
- [ ] https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/misc/jdiskreport/default.nix
- [ ] https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/perl-packages.nix
- [ ] https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/misc/adobe-reader/default.nix

**Comment by mibmo** [comment]
Most of them are use `rec` instead of the finalAttrs pattern (I'm not sure what it's formally called). Seems that they aren't run through a formatter either.

**Comment by mjarosie** [comment]
Additionally, GNU Hello package was deleted or moved, so the first link shows 404 error.

**Comment by rjpcasalino** [comment]
the unstable nixpkgs manual has moved the docs to a new [location](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md) the hello example is here now [`pkgs/by-name/he/hello/package.nix`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/he/hello/package.nix) - I assume the unstable nixpkgs manual rolls out later this month

**Comment by doronbehar** [comment]
I read today this part of our docs:

https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes

And it is not stated firmly whether using a function as an `mkDerivation` argument is better then the old `rec {}` option. I was wondering if that's intentional, and whether there are reasons not to use a function as an `mkDerivation` argument. I was pretty sure until I read this part that there is a strong consensus in favor of using `mkDerivation`... cc @roberth as the main pusher of the `finalAttrs:` pattern.

**Comment by doronbehar** [comment]
> > I was pretty sure until I read this part that there is a strong consensus in favor of using `mkDerivation`...
> 
> Don't both use `mkDerivation`? Could you clarify, @doronbehar ?

You are right, I wasn't clear, that "`mkderivation`..." should have been "`finalAttrs:`..." I edited that comment...

**Comment by mibmo** [comment]
> > Most of them are using `rec` instead of the `finalAttrs` pattern (I'm not sure what it's formally called), see
> 
> For what it's worth, I think part of the problem is that many builders still do not support explicitly-recursive attrsets.
> 
> #234651 introduces helpers which would make it much easier to support the pattern in all builders, FWIW

It's worth considering the eval time impact to nixpkgs as a whole too. I don't quite remember where, but I vaguely recall reading that a not-insignificant portion of the eval time is from derivations utilizing the `finalAttrs` pattern.
